This patch reworks the USB configuration for all the STM chips so that
instead of using #ifdefs based on the various chips, we pass in a flag
parameter to the wrapper configuration which describes the particular style
of configuration required for that chip.

This is primarily in preparation for the 7141, which has two different
styles of wrappers on the same chip.

Signed-off-by: Stuart Menefy <stuart.menefy@st.com>

Index: linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stb7100.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/kernel/cpu/sh4/setup-stb7100.c
+++ linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stb7100.c
@@ -79,23 +79,11 @@ struct platform_device wdt_device = {
 
 static u64 st40_dma_mask = 0xfffffff;
 
-static struct sysconf_field *usb_power_sc;
-
-static void usb_power_up(void* dev)
-{
-	unsigned long reg;
-
-	/* Make sure PLL is on */
-	reg = sysconf_read(usb_power_sc);
-	if (reg) {
-		sysconf_write(usb_power_sc, 0);
-		mdelay(100);
-	}
-}
-
 static struct plat_usb_data usb_wrapper =
 	USB_WRAPPER(0, AHB2STBUS_WRAPPER_GLUE_BASE, AHB2STBUS_PROTOCOL_BASE,
-		usb_power_up);
+		    USB_FLAGS_STRAP_16BIT	|
+		    USB_FLAGS_STRAP_PLL		|
+		    USB_FLAGS_OPC_MSGSIZE_CHUNKSIZE);
 
 static struct platform_device  st40_ohci_device =
 	USB_OHCI_DEVICE(0, AHB2STBUS_OHCI_BASE, 168, &usb_wrapper);
@@ -105,7 +93,9 @@ static struct platform_device  st40_ehci
 
 void __init stx7100_configure_usb(void)
 {
-	static struct stpio_pin *pin;
+	struct stpio_pin *pin;
+	struct sysconf_field *sc;
+	u32 reg;
 
 	/* Work around for USB over-current detection chip being
 	 * active low, and the 710x being active high.
@@ -145,7 +135,12 @@ void __init stx7100_configure_usb(void)
 	pin = stpio_request_pin(5,7, "USBPWR", STPIO_OUT);
 	stpio_set_pin(pin, 1);
 
-	usb_power_sc = sysconf_claim(SYS_CFG, 2, 1, 1, "usb");
+	sc = sysconf_claim(SYS_CFG, 2, 1, 1, "usb");
+	reg = sysconf_read(sc);
+	if (reg) {
+		sysconf_write(sc, 0);
+		mdelay(30);
+	}
 
 	platform_device_register(&st40_ohci_device);
 	platform_device_register(&st40_ehci_device);
Index: linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stx7105.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/kernel/cpu/sh4/setup-stx7105.c
+++ linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stx7105.c
@@ -75,9 +75,13 @@ static void stx7105_pio_sysconf(int bank
 
 static struct plat_usb_data usb_wrapper[2] = {
 	USB_WRAPPER(0, AHB2STBUS_WRAPPER_GLUE_BASE(0),
-		    AHB2STBUS_PROTOCOL_BASE(0), NULL),
+		    AHB2STBUS_PROTOCOL_BASE(0),
+		    USB_FLAGS_STRAP_8BIT	|
+		    USB_FLAGS_STBUS_CONFIG_THRESHOLD128),
 	USB_WRAPPER(1, AHB2STBUS_WRAPPER_GLUE_BASE(1),
-		    AHB2STBUS_PROTOCOL_BASE(1), NULL),
+		    AHB2STBUS_PROTOCOL_BASE(1),
+		    USB_FLAGS_STRAP_8BIT	|
+		    USB_FLAGS_STBUS_CONFIG_THRESHOLD128),
 };
 
 static struct platform_device st40_ohci_devices[2] = {
@@ -756,7 +760,7 @@ void __init stx7105_configure_asc(const 
 		struct platform_device *pdev;
 		struct stasc_uart_data *uart_data;
 
-		port = ascs[i];
+		port = ascs[i] & 0xff;
 		flags = ascs[i] >> 8;
 		pdev = &stm_stasc_devices[port];
 		uart_data = pdev->dev.platform_data;
Index: linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stx7111.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/kernel/cpu/sh4/setup-stx7111.c
+++ linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stx7111.c
@@ -40,7 +40,9 @@ static u64 st40_dma_mask = 0xfffffff;
 
 static struct plat_usb_data usb_wrapper =
 	USB_WRAPPER(0, AHB2STBUS_WRAPPER_GLUE_BASE, AHB2STBUS_PROTOCOL_BASE,
-		NULL);
+		    USB_FLAGS_STRAP_16BIT	|
+		    USB_FLAGS_STRAP_PLL		|
+		    USB_FLAGS_STBUS_CONFIG_THRESHOLD256);
 
 static struct platform_device  st40_ohci_device =
 	USB_OHCI_DEVICE(0, AHB2STBUS_OHCI_BASE,
@@ -52,7 +54,7 @@ static struct platform_device  st40_ehci
 			evt2irq(0x1720), /* 169 */
 			&usb_wrapper);
 
-void __init stx7111_configure_usb(void)
+void __init stx7111_configure_usb(int inv_enable)
 {
 	static struct stpio_pin *pin;
 	struct sysconf_field *sc;
@@ -67,7 +69,7 @@ void __init stx7111_configure_usb(void)
 	 * an inverter on the overcurrent signal.
 	 */
 	sc = sysconf_claim(SYS_CFG, 6, 29,29, "USB");
-	sysconf_write(sc, 1);
+	sysconf_write(sc, inv_enable);
 
 	pin = stpio_request_pin(5,6, "USBOC", STPIO_IN);
 	pin = stpio_request_pin(5,7, "USBPWR", STPIO_ALT_OUT);
@@ -447,7 +449,7 @@ void __init stx7111_configure_asc(const 
 		struct platform_device *pdev;
 		struct sysconf_field *sc;
 
-		port = ascs[i];
+		port = ascs[i] & 0xff;
 		flags = ascs[i] >> 8;
 		pdev = &stm_stasc_devices[port];
 
Index: linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stx7200.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/kernel/cpu/sh4/setup-stx7200.c
+++ linux-2.6.23-stm/arch/sh/kernel/cpu/sh4/setup-stx7200.c
@@ -43,24 +43,22 @@ static struct sysconf_field *sc7_2;
 
 static u64 st40_dma_mask = 0xfffffff;
 
-static struct sysconf_field *usb_power_sc[3];
-
-static void usb_power_up(void* dev)
-{
-	struct platform_device *pdev = dev;
-	struct plat_usb_data *usb_wrapper = pdev->dev.platform_data;
-	int port = usb_wrapper->port_number;
-
-	sysconf_write(usb_power_sc[port], 0);
-}
-
 static struct plat_usb_data usb_wrapper[3] = {
 	USB_WRAPPER(0, AHB2STBUS_WRAPPER_GLUE_BASE(0),
-		    AHB2STBUS_PROTOCOL_BASE(0), usb_power_up),
+		    AHB2STBUS_PROTOCOL_BASE(0),
+		    USB_FLAGS_STRAP_8BIT		|
+		    USB_FLAGS_STRAP_PLL			|
+		    USB_FLAGS_OPC_MSGSIZE_CHUNKSIZE),
 	USB_WRAPPER(1, AHB2STBUS_WRAPPER_GLUE_BASE(1),
-		    AHB2STBUS_PROTOCOL_BASE(1), usb_power_up),
+		    AHB2STBUS_PROTOCOL_BASE(1),
+		    USB_FLAGS_STRAP_8BIT		|
+		    USB_FLAGS_STRAP_PLL			|
+		    USB_FLAGS_OPC_MSGSIZE_CHUNKSIZE),
 	USB_WRAPPER(2, AHB2STBUS_WRAPPER_GLUE_BASE(2),
-		    AHB2STBUS_PROTOCOL_BASE(2), usb_power_up),
+		    AHB2STBUS_PROTOCOL_BASE(2),
+		    USB_FLAGS_STRAP_8BIT		|
+		    USB_FLAGS_STRAP_PLL			|
+		    USB_FLAGS_OPC_MSGSIZE_CHUNKSIZE),
 };
 
 static struct platform_device st40_ohci_devices[3] = {
@@ -336,27 +334,26 @@ static void __init usb_soft_jtag_reset(v
 	sysconf_write(sc, 0x00000040);
 }
 
-void __init stx7200_configure_usb(void)
+void __init stx7200_configure_usb(int port)
 {
+	static int first = 1;
 	const unsigned char power_pins[3] = {1, 3, 4};
 	const unsigned char oc_pins[3] = {0, 2, 5};
-	static struct stpio_pin *pio;
+	struct stpio_pin *pio;
 	struct sysconf_field *sc;
-	int port;
 
-	/* route USB and parts of MAFE instead of DVO.
-	 * conf_pad_pio[2] = 0 */
-	sc = sysconf_claim(SYS_CFG, 7, 26, 26, "usb");
-	sysconf_write(sc, 0);
+	if (first) {
+		/* route USB and parts of MAFE instead of DVO.
+		 * conf_pad_pio[2] = 0 */
+		sc = sysconf_claim(SYS_CFG, 7, 26, 26, "usb");
+		sysconf_write(sc, 0);
 
-	/* DVO output selection (probably ignored).
-	 * conf_pad_pio[3] = 0 */
-	sc = sysconf_claim(SYS_CFG, 7, 27, 27, "usb");
-	sysconf_write(sc, 0);
+		/* DVO output selection (probably ignored).
+		 * conf_pad_pio[3] = 0 */
+		sc = sysconf_claim(SYS_CFG, 7, 27, 27, "usb");
+		sysconf_write(sc, 0);
 
-	if (cpu_data->cut_major < 2) {
 		/* Enable soft JTAG mode for USB and SATA
-		 * Taken from OS21, but is this correct?
 		 * soft_jtag_en = 1 */
 		sc = sysconf_claim(SYS_CFG, 33, 6, 6, "usb");
 		sysconf_write(sc, 1);
@@ -364,27 +361,29 @@ void __init stx7200_configure_usb(void)
 		sc = sysconf_claim(SYS_CFG, 33, 0, 3, "usb");
 		sysconf_write(sc, 0);
 
-		usb_soft_jtag_reset();
+		if (cpu_data->cut_major < 2)
+			usb_soft_jtag_reset();
+
+		first = 0;
 	}
 
-	for (port=0; port<3; port++) {
-		usb_power_sc[port] = sysconf_claim(SYS_CFG, 22, 3+port,
-						   3+port, "usb");
+	/* Power up port */
+	sc = sysconf_claim(SYS_CFG, 22, 3+port, 3+port, "usb");
+	sysconf_write(sc, 0);
 
-		pio = stpio_request_pin(7, power_pins[port], "USB power",
-					STPIO_ALT_OUT);
-		stpio_set_pin(pio, 1);
+	pio = stpio_request_pin(7, power_pins[port], "USB power",
+				STPIO_ALT_OUT);
+	stpio_set_pin(pio, 1);
 
-		if (cpu_data->cut_major < 2)
-			pio = stpio_request_pin(7, oc_pins[port], "USB oc",
+	if (cpu_data->cut_major < 2)
+		pio = stpio_request_pin(7, oc_pins[port], "USB oc",
 					STPIO_ALT_BIDIR);
-		else
-			pio = stpio_request_pin(7, oc_pins[port], "USB oc",
+	else
+		pio = stpio_request_pin(7, oc_pins[port], "USB oc",
 					STPIO_IN);
 
-		platform_device_register(&st40_ohci_devices[port]);
-		platform_device_register(&st40_ehci_devices[port]);
-	}
+	platform_device_register(&st40_ohci_devices[port]);
+	platform_device_register(&st40_ehci_devices[port]);
 }
 
 /* SATA resources ---------------------------------------------------------- */
Index: linux-2.6.23-stm/drivers/usb/Kconfig
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/Kconfig
+++ linux-2.6.23-stm/drivers/usb/Kconfig
@@ -45,6 +45,7 @@ config USB_ARCH_HAS_OHCI
 	default y if CPU_SUBTYPE_STB7100
 	default y if CPU_SUBTYPE_STX7105
 	default y if CPU_SUBTYPE_STX7111
+	default y if CPU_SUBTYPE_STX7141
 	default y if CPU_SUBTYPE_STX7200
 	# more:
 	default PCI
@@ -57,6 +58,7 @@ config USB_ARCH_HAS_EHCI
 	default y if CPU_SUBTYPE_STB7100
 	default y if CPU_SUBTYPE_STX7105
 	default y if CPU_SUBTYPE_STX7111
+	default y if CPU_SUBTYPE_STX7141
 	default y if CPU_SUBTYPE_STX7200
 	default PCI
 
Index: linux-2.6.23-stm/drivers/usb/host/ehci-hcd.c
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/host/ehci-hcd.c
+++ linux-2.6.23-stm/drivers/usb/host/ehci-hcd.c
@@ -953,6 +953,7 @@ MODULE_LICENSE ("GPL");
 #if defined(CONFIG_CPU_SUBTYPE_STB7100) || \
     defined(CONFIG_CPU_SUBTYPE_STX7105) || \
     defined(CONFIG_CPU_SUBTYPE_STX7111) || \
+    defined(CONFIG_CPU_SUBTYPE_STX7141) || \
     defined(CONFIG_CPU_SUBTYPE_STX7200)
 #include "ehci-stcore.c"
 #define	PLATFORM_DRIVER		ehci_hcd_st40_driver
Index: linux-2.6.23-stm/drivers/usb/host/ehci-stcore.c
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/host/ehci-stcore.c
+++ linux-2.6.23-stm/drivers/usb/host/ehci-stcore.c
@@ -26,16 +26,6 @@ static int ehci_st40_resume(struct usb_h
 }
 #endif
 
-static irqreturn_t ehci_st40_irq(struct usb_hcd *hcd)
-{
-	irqreturn_t retval;
-
-	usb_hcd_st40_wait_irq();
-	retval = ehci_irq(hcd);
-
-	return retval;
-}
-
 static int ehci_st40_reset(struct usb_hcd *hcd)
 {
 	writel(AHB2STBUS_INOUT_THRESHOLD,
@@ -45,13 +35,13 @@ static int ehci_st40_reset(struct usb_hc
 
 static const struct hc_driver ehci_st40_hc_driver = {
 	.description = hcd_name,
-	.product_desc = "ST EHCI Host Controller",
+	.product_desc = "STM EHCI Host Controller",
 	.hcd_priv_size = sizeof(struct ehci_hcd),
 
 	/*
 	 * generic hardware linkage
 	 */
-	.irq = ehci_st40_irq,
+	.irq = ehci_irq,
 	.flags = HCD_MEMORY | HCD_USB2,
 
 	/*
@@ -106,7 +96,7 @@ static int ehci_hcd_st40_probe(const str
 	if (retval)
 		return retval;
 
-	hcd = usb_create_hcd(driver, &dev->dev, "STB7100_EHCI");
+	hcd = usb_create_hcd(driver, &dev->dev, dev->dev.bus_id);
 	if (!hcd) {
 		retval = -ENOMEM;
 		goto err0;
@@ -135,7 +125,7 @@ static int ehci_hcd_st40_probe(const str
 	/* cache this readonly data; minimize device reads */
 	ehci->hcs_params = readl(&ehci->caps->hcs_params);
 
-	retval=usb_add_hcd(hcd, dev->resource[1].start, SA_SHIRQ);
+	retval = usb_add_hcd(hcd, dev->resource[1].start, 0);
 	if (retval == 0)
 		return retval;
 
Index: linux-2.6.23-stm/drivers/usb/host/ohci-hcd.c
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/host/ohci-hcd.c
+++ linux-2.6.23-stm/drivers/usb/host/ohci-hcd.c
@@ -918,6 +918,7 @@ MODULE_LICENSE ("GPL");
 #if defined(CONFIG_CPU_SUBTYPE_STB7100) || \
 	defined(CONFIG_CPU_SUBTYPE_STX7105) || \
 	defined(CONFIG_CPU_SUBTYPE_STX7111) || \
+	defined(CONFIG_CPU_SUBTYPE_STX7141) || \
 	defined(CONFIG_CPU_SUBTYPE_STX7200)
 #include "ohci-stcore.c"
 #define PLATFORM_DRIVER		ohci_hcd_stm_driver
Index: linux-2.6.23-stm/drivers/usb/host/ohci-stcore.c
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/host/ohci-stcore.c
+++ linux-2.6.23-stm/drivers/usb/host/ohci-stcore.c
@@ -16,19 +16,8 @@
 
 extern int usb_disabled(void);
 
-#define DEVICE_NAME "STB7100 OHCI"
 #include "stb7100-common.h"
 
-static irqreturn_t ohci_st40_irq(struct usb_hcd *hcd)
-{
-	irqreturn_t retval;
-
-	usb_hcd_st40_wait_irq();
-	retval = ohci_irq(hcd);
-
-	return retval;
-}
-
 static int
 ohci_st40_start(struct usb_hcd *hcd)
 {
@@ -47,6 +36,7 @@ ohci_st40_start(struct usb_hcd *hcd)
 	return 0;
 }
 
+#ifdef	CONFIG_PM
 static int
 ohci_st40_suspend(struct usb_hcd *hcd, pm_message_t message)
 {
@@ -58,14 +48,15 @@ ohci_st40_resume(struct usb_hcd *hcd)
 {
 	return 0;
 }
+#endif
 
 static const struct hc_driver ohci_st40_hc_driver = {
 	.description =		hcd_name,
-	.product_desc =		DEVICE_NAME,
+	.product_desc =		"STM OHCI Host Controller",
 	.hcd_priv_size =	sizeof(struct ohci_hcd),
 
 	/* generic hardware linkage */
-	.irq =			ohci_st40_irq,
+	.irq =			ohci_irq,
 	.flags =		HCD_USB11 | HCD_MEMORY,
 
 	/* basic lifecycle operations */
@@ -108,7 +99,7 @@ static int ohci_hcd_stm_probe(struct pla
 	if (retval)
 		return retval;
 
-	hcd = usb_create_hcd(driver, &pdev->dev, DEVICE_NAME);
+	hcd = usb_create_hcd(driver, &pdev->dev, pdev->dev.bus_id);
 	if (!hcd) {
 		pr_debug("hcd_create_hcd failed");
 		retval = -ENOMEM;
@@ -133,7 +124,7 @@ static int ohci_hcd_stm_probe(struct pla
 
 	ohci_hcd_init(hcd_to_ohci(hcd));
 
-	retval = usb_add_hcd(hcd, pdev->resource[1].start, SA_INTERRUPT);
+	retval = usb_add_hcd(hcd, pdev->resource[1].start, 0);
 	if (retval == 0)
 		return retval;
 
Index: linux-2.6.23-stm/drivers/usb/host/stb7100-common.c
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/host/stb7100-common.c
+++ linux-2.6.23-stm/drivers/usb/host/stb7100-common.c
@@ -14,16 +14,6 @@
 #include <asm/io.h>
 #include "stb7100-common.h"
 
-#ifdef CONFIG_CPU_SUBTYPE_STX7200
-#define STRAP_MODE	0 /* 8 bit */
-#define MSGSIZE		AHB2STBUS_MSGSIZE_4
-#define CHUNKSIZE	AHB2STBUS_CHUNKSIZE_4
-#else
-#define STRAP_MODE	AHB2STBUS_STRAP_16_BIT
-#define MSGSIZE		AHB2STBUS_MSGSIZE_64
-#define CHUNKSIZE	AHB2STBUS_CHUNKSIZE_64
-#endif
-
 #define RESOURCE_NAME "USB wrapper"
 
 static DEFINE_MUTEX(wraper_mutex);
@@ -68,74 +58,70 @@ int ST40_start_host_control(struct platf
 	if (!protocol_base)
 		goto err4;
 
-
-#if	defined(CONFIG_CPU_SUBTYPE_STX7105)
-	req_reg =
-		(1<<21) |  /* Turn on read-ahead */
-		(5<<16) |  /* Opcode is store/load 32 */
-		(0<<15) |  /* Turn off write posting */
-		(1<<14) |  /* Enable threshold */
-		(3<<9)  |  /* 2**3 Packets in a chunk */
-		(0<<4)  |  /* No messages */
-		7;         /* Threshold is 128 */
-
-	do {
-		writel(req_reg, protocol_base + AHB2STBUS_STBUS_CONFIG);
-		reg = readl(protocol_base + AHB2STBUS_STBUS_CONFIG);
-	} while ((reg & 0x7FFFFFFF) != req_reg);
-#else
-	/* Set strap mode */
-	reg = readl(wrapper_base + AHB2STBUS_STRAP_OFFSET);
-	reg &= ~AHB2STBUS_STRAP_16_BIT;
-	reg |= STRAP_MODE;
-	writel(reg, wrapper_base + AHB2STBUS_STRAP_OFFSET);
-
-	/* Start PLL */
-	reg = readl(wrapper_base + AHB2STBUS_STRAP_OFFSET);
-	writel(reg | AHB2STBUS_STRAP_PLL,
-	       wrapper_base + AHB2STBUS_STRAP_OFFSET);
-	mdelay(100);
-	writel(reg & (~AHB2STBUS_STRAP_PLL),
-	       wrapper_base + AHB2STBUS_STRAP_OFFSET);
-	mdelay(100);
-
-#if	defined(CONFIG_CPU_SUBTYPE_STB7100) || \
-	defined(CONFIG_CPU_SUBTYPE_STX7200)
-
-	/* Set the STBus Opcode Config for load/store 32 */
-	writel(AHB2STBUS_STBUS_OPC_32BIT,
-	       protocol_base + AHB2STBUS_STBUS_OPC_OFFSET);
-
-	/* Set the Message Size Config to n packets per message */
-	writel(MSGSIZE,
-	       protocol_base + AHB2STBUS_MSGSIZE_OFFSET);
-
-	writel(CHUNKSIZE,
-	       protocol_base + AHB2STBUS_CHUNKSIZE_OFFSET);
-
-#elif	defined(CONFIG_CPU_SUBTYPE_STX7111)
-
-	req_reg =
-		(1<<21) |  /* Turn on read-ahead */
-		(5<<16) |  /* Opcode is store/load 32 */
-		(0<<15) |  /* Turn off write posting */
-		(1<<14) |  /* Enable threshold */
-		(3<<9)  |  /* 2**3 Packets in a chunk */
-		(0<<4)  |  /* No messages */
-		(8<<0);    /* Threshold is 256 */
-
-	do {
-		writel(req_reg, protocol_base + AHB2STBUS_STBUS_CONFIG);
-		reg = readl(protocol_base + AHB2STBUS_STBUS_CONFIG);
-	} while ((reg & 0x7FFFFFFF) != req_reg);
-
-#else
-#error Unknown CPU
-#endif
-#endif
-
-	if (usb_wrapper->power_up)
-		usb_wrapper->power_up(pdev);
+	if (usb_wrapper->flags &
+	    (USB_FLAGS_STRAP_8BIT | USB_FLAGS_STRAP_16BIT)) {
+		/* Set strap mode */
+		reg = readl(wrapper_base + AHB2STBUS_STRAP_OFFSET);
+		if (usb_wrapper->flags & USB_FLAGS_STRAP_16BIT)
+			reg |= AHB2STBUS_STRAP_16_BIT;
+		else
+			reg &= ~AHB2STBUS_STRAP_16_BIT;
+		writel(reg, wrapper_base + AHB2STBUS_STRAP_OFFSET);
+	}
+
+	if (usb_wrapper->flags & USB_FLAGS_STRAP_PLL) {
+		/* Start PLL */
+		reg = readl(wrapper_base + AHB2STBUS_STRAP_OFFSET);
+		writel(reg | AHB2STBUS_STRAP_PLL,
+		       wrapper_base + AHB2STBUS_STRAP_OFFSET);
+		mdelay(30);
+		writel(reg & (~AHB2STBUS_STRAP_PLL),
+		       wrapper_base + AHB2STBUS_STRAP_OFFSET);
+		mdelay(30);
+	}
+
+	if (usb_wrapper->flags & USB_FLAGS_OPC_MSGSIZE_CHUNKSIZE) {
+		/* Set the STBus Opcode Config for load/store 32 */
+		writel(AHB2STBUS_STBUS_OPC_32BIT,
+		       protocol_base + AHB2STBUS_STBUS_OPC_OFFSET);
+
+		/* Set the Message Size Config to n packets per message */
+		writel(AHB2STBUS_MSGSIZE_4,
+		       protocol_base + AHB2STBUS_MSGSIZE_OFFSET);
+
+		/* Set the chunksize to n packets */
+		writel(AHB2STBUS_CHUNKSIZE_4,
+		       protocol_base + AHB2STBUS_CHUNKSIZE_OFFSET);
+	}
+
+	if (usb_wrapper->flags &
+	    (USB_FLAGS_STBUS_CONFIG_THRESHOLD128 |
+	     USB_FLAGS_STBUS_CONFIG_THRESHOLD256)) {
+
+		if (usb_wrapper->flags & USB_FLAGS_STBUS_CONFIG_THRESHOLD128)
+			req_reg =
+				(1<<21) |  /* Turn on read-ahead */
+				(5<<16) |  /* Opcode is store/load 32 */
+				(0<<15) |  /* Turn off write posting */
+				(1<<14) |  /* Enable threshold */
+				(3<<9)  |  /* 2**3 Packets in a chunk */
+				(0<<4)  |  /* No messages */
+				7;         /* Threshold is 128 */
+		else
+			req_reg =
+				(1<<21) |  /* Turn on read-ahead */
+				(5<<16) |  /* Opcode is store/load 32 */
+				(0<<15) |  /* Turn off write posting */
+				(1<<14) |  /* Enable threshold */
+				(3<<9)  |  /* 2**3 Packets in a chunk */
+				(0<<4)  |  /* No messages */
+				(8<<0);    /* Threshold is 256 */
+
+		do {
+			writel(req_reg, protocol_base + AHB2STBUS_STBUS_CONFIG);
+			reg = readl(protocol_base + AHB2STBUS_STBUS_CONFIG);
+		} while ((reg & 0x7FFFFFFF) != req_reg);
+	}
 
 	usb_wrapper->initialised = 1;
 
Index: linux-2.6.23-stm/drivers/usb/host/stb7100-common.h
===================================================================
--- linux-2.6.23-stm.orig/drivers/usb/host/stb7100-common.h
+++ linux-2.6.23-stm/drivers/usb/host/stb7100-common.h
@@ -8,11 +8,13 @@
  */
 
 /*
- * Some of these register are described in ADCS 7518758.
+ * This file attempts to support all the various flavours of USB wrappers,
+ * thus some of the registers appear to overlap.
+ *
+ * Some of these register are described in ADCS 7518758 and 7618754
  */
 
-#if	defined(CONFIG_CPU_SUBTYPE_STB7100) || \
-	defined(CONFIG_CPU_SUBTYPE_STX7200)
+/* Protocol converter registers (separate registers) */
 
 /* The transaction opcode is programmed in this register */
 #define AHB2STBUS_STBUS_OPC_OFFSET	0x00	/* From AHB2STBUS_PROTOCOL_BASE */
@@ -42,77 +44,25 @@
 #define AHB2STBUS_CHUNKSIZE_32		0x5
 #define AHB2STBUS_CHUNKSIZE_64		0x6
 
-/* This register holds the timeout value in number of STBus clock cycles */
-#define AHB2STBUS_REQ_TIMEOUT		(AHB2STBUS_PROTOCOL_BASE + 0x0c)
 
-/* Undocumented */
-#define AHB2STBUS_PC_STATUS		(AHB2STBUS_PROTOCOL_BASE + 0x10)
-#define AHB2STBUS_PC_STATUS_IDLE	1
+/* Protocol converter registers (combined register) */
 
-#elif	defined(CONFIG_CPU_SUBTYPE_STX7105) || \
-	defined(CONFIG_CPU_SUBTYPE_STX7111)
-
-/* No documentation for this */
 #define AHB2STBUS_STBUS_CONFIG		0x04	/* From AHB2STBUS_PROTOCOL_BASE */
 
-#else
-#error Unknown CPU
-#endif
-
-/* This register implements interrupt status for the OHCI controller */
-#define AHB2STBUS_OHCI_INT_STS		(AHB2STBUS_WRAPPER_GLUE_BASE + 0x08)
-
-/* This register implements interrupt mask for the OHCI controller */
-#define AHB2STBUS_OHCI_INT_MASK		(AHB2STBUS_WRAPPER_GLUE_BASE + 0x0c)
-
-/* This register implements interrupt status for the EHCI controller */
-#define AHB2STBUS_EHCI_INT_STS		(AHB2STBUS_WRAPPER_GLUE_BASE + 0x10)
 
+/* Wrapper Glue registers */
 
 #define AHB2STBUS_STRAP_OFFSET		0x14	/* From AHB2STBUS_WRAPPER_GLUE_BASE */
 #define AHB2STBUS_STRAP_PLL		0x08	/* undocumented */
+#define AHB2STBUS_STRAP_8_BIT		0x00	/* ss_word_if */
 #define AHB2STBUS_STRAP_16_BIT		0x04	/* ss_word_if */
 
+
+/* Extensions to the standard USB register set */
+
 /* Define a bus wrapper IN/OUT threshold of 128 */
 #define AHB2STBUS_INSREG01_OFFSET	(0x10 + 0x84) /* From AHB2STBUS_EHCI_BASE */
 #define AHB2STBUS_INOUT_THRESHOLD	0x00800080
 
-static inline void usb_hcd_st40_wait_irq(void)
-{
-#if 0
-	/*
-	 * Fix required to work around a problem which causes controller
-	 * memory writes to be overtaken by interrupt requests.
-	 *
-	 * From the document:
-	 * STBus USB Host 2.0 Controller
-	 * Known Problems and Workaround
-	 *
-	 * 2.1 Interrupt Generation not linked with completed read/write
-	 * on STBUS
-	 *
-	 * This limitation occurs because any writes issued by the AHB
-	 * Master of the Synopsys Controller to the AHB Slave of the
-	 * protocol converter is acknowledged by the STBUS target
-	 * immediately to the AHB Master, before the transfer is
-	 * completed by the STBUS Inititator.This causes the OHCI
-	 * Interrupt to be generated before the transfer is completed
-	 * on the STBUS. This may cause problems when the OHCI ISR
-	 * successfully reads the memory location before the completion
-	 * of the write by the STBUS Initiator of the DUT.
-	 *
-	 * This problems has also been observed in the EHCI controller.
-	 */
-	int count = 0;
-
-	while ((readl(AHB2STBUS_PC_STATUS) & AHB2STBUS_PC_STATUS_IDLE) == 0) {
-		count++;
-		if (count == 100) {
-			warn("OHCI AHB interrupt sync looped too many times");
-			break;
-		}
-	}
-#endif
-}
 
 int ST40_start_host_control(struct platform_device *dev);
Index: linux-2.6.23-stm/include/linux/stm/soc.h
===================================================================
--- linux-2.6.23-stm.orig/include/linux/stm/soc.h
+++ linux-2.6.23-stm/include/linux/stm/soc.h
@@ -166,11 +166,18 @@ struct plat_stmmacphy_data {
 struct plat_usb_data {
 	unsigned long ahb2stbus_wrapper_glue_base;
 	unsigned long ahb2stbus_protocol_base;
-	void (*power_up)(void* dev);
+	unsigned long flags;
 	int initialised;
 	int port_number;
 };
 
+#define USB_FLAGS_STRAP_8BIT			(1<<0)
+#define USB_FLAGS_STRAP_16BIT			(2<<0)
+#define USB_FLAGS_STRAP_PLL			(1<<2)
+#define USB_FLAGS_OPC_MSGSIZE_CHUNKSIZE		(1<<3)
+#define USB_FLAGS_STBUS_CONFIG_THRESHOLD128	(1<<4)
+#define USB_FLAGS_STBUS_CONFIG_THRESHOLD256	(2<<4)
+
 /**
  * struct usb_init_data - initialisation data for a USB port
  * @oc_en: enable OC detection (0 or 1)
@@ -263,7 +270,7 @@ void stx7111_early_device_init(void);
 void stx7111_configure_asc(const int *ascs, int num_ascs, int console);
 void stx7111_configure_pwm(struct plat_stm_pwm_data *data);
 void stx7111_configure_ssc(struct plat_ssc_data *data);
-void stx7111_configure_usb(void);
+void stx7111_configure_usb(int inv_enable);
 void stx7111_configure_ethernet(int en_mii, int sel, int ext_clk, int phy_bus);
 void stx7111_configure_nand(struct nand_config_data *data);
 void stx7111_configure_lirc(lirc_scd_t *scd);
@@ -272,7 +279,7 @@ void stx7200_early_device_init(void);
 void stx7200_configure_asc(const int *ascs, int num_ascs, int console);
 void stx7200_configure_pwm(struct plat_stm_pwm_data *data);
 void stx7200_configure_ssc(struct plat_ssc_data *data);
-void stx7200_configure_usb(void);
+void stx7200_configure_usb(int port);
 void stx7200_configure_sata(unsigned int port);
 void stx7200_configure_ethernet(int mac, int rmii_mode, int ext_clk,
 				int phy_bus);
Index: linux-2.6.23-stm/include/linux/stm/soc_init.h
===================================================================
--- linux-2.6.23-stm.orig/include/linux/stm/soc_init.h
+++ linux-2.6.23-stm/include/linux/stm/soc_init.h
@@ -73,11 +73,11 @@
 	}								\
 }
 
-#define USB_WRAPPER(_port, _wrapper_base, _protocol_base, _power_up)	\
+#define USB_WRAPPER(_port, _wrapper_base, _protocol_base, _flags)	\
 {									\
 	.ahb2stbus_wrapper_glue_base = _wrapper_base,			\
 	.ahb2stbus_protocol_base = _protocol_base,			\
-	.power_up = _power_up,						\
+	.flags = _flags,						\
 	.initialised = 0,						\
 	.port_number = _port,						\
 }
Index: linux-2.6.23-stm/arch/sh/boards/st/cb101/setup.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/boards/st/cb101/setup.c
+++ linux-2.6.23-stm/arch/sh/boards/st/cb101/setup.c
@@ -210,7 +210,9 @@ static struct platform_device *cb101_dev
 static int __init device_init(void)
 {
 	stx7200_configure_ssc(&ssc_private_info);
-	stx7200_configure_usb();
+	stx7200_configure_usb(0);
+	stx7200_configure_usb(1);
+	stx7200_configure_usb(2);
 	stx7200_configure_ethernet(0, 0, 0, 0);
 	stx7200_configure_lirc(NULL);
 	stx7200_configure_nand(&cb101_nand_config[0]);
Index: linux-2.6.23-stm/arch/sh/boards/st/cb102/setup.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/boards/st/cb102/setup.c
+++ linux-2.6.23-stm/arch/sh/boards/st/cb102/setup.c
@@ -168,7 +168,9 @@ static int __init device_init(void)
 	BUG_ON(pio == NULL);
 	stpio_set_pin(pio, 1);
 
-	stx7200_configure_usb();
+	stx7200_configure_usb(0);
+	stx7200_configure_usb(1);
+	stx7200_configure_usb(2);
 
 	stx7200_configure_ethernet(0, 0, 0, 0);
 
Index: linux-2.6.23-stm/arch/sh/boards/st/mb519/setup.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/boards/st/mb519/setup.c
+++ linux-2.6.23-stm/arch/sh/boards/st/mb519/setup.c
@@ -203,7 +203,9 @@ static int __init device_init(void)
 
 	stx7200_configure_pwm(&pwm_private_info);
 	stx7200_configure_ssc(&ssc_private_info);
-	stx7200_configure_usb();
+	stx7200_configure_usb(0);
+	stx7200_configure_usb(1);
+	stx7200_configure_usb(2);
 #if 1 /* On-board PHY (MII0) */
 	stx7200_configure_ethernet(0, 0, 1, 0);
 #else /* External PHY board (MII1) */
Index: linux-2.6.23-stm/arch/sh/boards/st/mb671/setup.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/boards/st/mb671/setup.c
+++ linux-2.6.23-stm/arch/sh/boards/st/mb671/setup.c
@@ -193,7 +193,9 @@ static int __init device_init(void)
 	stx7200_configure_pwm(&pwm_private_info);
 	stx7200_configure_ssc(&ssc_private_info);
 
-	stx7200_configure_usb();
+	stx7200_configure_usb(0);
+	stx7200_configure_usb(1);
+	stx7200_configure_usb(2);
 
 	stx7200_configure_sata(0);
 
Index: linux-2.6.23-stm/arch/sh/boards/st/mb618/setup.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/boards/st/mb618/setup.c
+++ linux-2.6.23-stm/arch/sh/boards/st/mb618/setup.c
@@ -269,7 +269,7 @@ static int __init device_init(void)
 {
 	stx7111_configure_pwm(&pwm_private_info);
 	stx7111_configure_ssc(&ssc_private_info);
-	stx7111_configure_usb();
+	stx7111_configure_usb(1); /* Enable inverter */
 	stx7111_configure_ethernet(1, 0, 0, 0);
         stx7111_configure_lirc(&lirc_scd);
 
